packages/fx-core/src/component/generator/utils.ts

```diff
@@ -0,0 +1,181 @@
+  timeoutInMs = defaultTimeoutInMs
+): Promise<string> {
+  const tags = await fetchTemplateTagList(templateTagListUrl, tryLimits, timeoutInMs);
+  const selectedTag = selectTag(tags.replace(/\r/g, "").split("\n"));
+  if (!selectedTag) {
+    throw new Error(`Failed to find valid template for ${name}`);
+  }
+  return `${baseUrl}/${selectTag}/${templateZipName(name)}`;
+}
+
+export const templateZipName = (templateName: string): string => `${templateName}.zip`;
+
+export function renderTemplateFileData(
+  fileName: string,
+  fileData: Buffer,
+  variables?: { [key: string]: string }
+): string | Buffer {
+  //only mustache files with name ending with .tql
+  if (path.extname(fileName) === templateFileExt) {
+    return Mustache.render(fileData.toString(), variables);
+  }
+  // Return Buffer instead of string if the file is not a template. Because `toString()` may break binary resources, like png files.
+  return fileData;
+}
+
+export function renderTemplateFileName(
+  fileName: string,
+  fileData: Buffer,
+  variables?: { [key: string]: string }
+): string {
+  return Mustache.render(fileName, variables);
+}
+
+export function genFileDataRenderReplaceFn(variables: { [key: string]: string }) {
+  return (fileName: string, fileData: Buffer) =>
+    renderTemplateFileData(fileName, fileData, variables);
+}
+
+export function genFileNameRenderReplaceFn(variables: { [key: string]: string }) {
+  return (fileName: string, fileData: Buffer) =>
+    renderTemplateFileName(fileName, fileData, variables).replace(templateFileExt, "");
+}
+
+//the unzip function does the following things:
+//1. unzip the package into dstPath,
+//2. replace the file name and file content with the given replace functions
+//3. if appFolder is provided, only the files within appFolder will be kept. This is used for samples from other repos.
+export async function unzip(
+  zip: AdmZip,
+  dstPath: string,
+  relativePath?: string,
+  nameReplaceFn?: (filePath: string, data: Buffer) => string,
+  dataReplaceFn?: (filePath: string, data: Buffer) => Buffer | string
+): Promise<void> {
+  let entries: AdmZip.IZipEntry[] = zip.getEntries().filter((entry) => !entry.isDirectory);
+  if (relativePath) {
+    entries = entries.filter((entry) => entry.entryName.startsWith(relativePath));
+  }
+
+  for (const entry of entries) {
+    const rawEntryData: Buffer = entry.getData();
+    let entryName: string = nameReplaceFn
+      ? nameReplaceFn(entry.entryName, rawEntryData)
+      : entry.entryName;
+    if (relativePath) {
+      entryName = entryName.replace(relativePath, "");
+    }
+    const entryData: string | Buffer = dataReplaceFn
+      ? dataReplaceFn(entry.name, rawEntryData)
+      : rawEntryData;
+
+    const filePath: string = path.join(dstPath, entryName);
+    const dirPath: string = path.dirname(filePath);
+    await fs.ensureDir(dirPath);
+    await fs.writeFile(filePath, entryData);
+  }
+}
+
+export async function getValidSampleDestination(
```

Comment:
although we may copy this code part from sample gallery, since it will be code of Generator, we could write it in a better way with high code quality.